File: packages/sdk-ts/src/resources/leads.ts
import type { Http, RequestOptions } from '../core/http.js';
import type { Page } from '../core/page.js';
import type { BulkCreateLeads, BulkLeadResult, CreateLead, Lead, LeadListQuery, UpdateLead } from '../types.js';

const enc = encodeURIComponent;

/** Leads: the contacts agents call. */
export class Leads {
    constructor(private readonly http: Http) {}

    list(query: LeadListQuery = {}, opts?: RequestOptions): Promise<Page<Lead>> {
        return this.http.list<Lead>('/v1/leads', query, opts);
    }

    create(body: CreateLead, opts?: RequestOptions): Promise<Lead> {
        return this.http.request<Lead>('POST', '/v1/leads', { ...opts, body });
    }

    /** Up to 1,000 at once. Rows are validated individually; check `failed_count` and `errors`. */
    bulkCreate(body: BulkCreateLeads, opts?: RequestOptions): Promise<BulkLeadResult> {
        return this.http.request<BulkLeadResult>('POST', '/v1/leads/bulk', { ...opts, body });
    }

    retrieve(id: string, opts?: RequestOptions): Promise<Lead> {
        return this.http.request<Lead>('GET', `/v1/leads/${enc(id)}`, opts);
    }

    /** Only the fields sent are changed. */
    update(id: string, body: UpdateLead, opts?: RequestOptions): Promise<Lead> {
        return this.http.request<Lead>('PATCH', `/v1/leads/${enc(id)}`, { ...opts, body });
    }

    delete(id: string, opts?: RequestOptions): Promise<void> {
        return this.http.request<void>('DELETE', `/v1/leads/${enc(id)}`, opts);
    }
}

File: packages/sdk-dart/lib/src/resources/leads.dart
import '../models/lead.dart';
import '../page.dart';
import '../transport.dart';

/// Leads: the contacts agents call.
class Leads {
  /// Binds the resource to a transport.
  const Leads(this._transport);

  final Transport _transport;

  /// Lists leads. Filter by tag or free-text search.
  Future<Page<Lead>> list({
    int? limit,
    String? startingAfter,
    String? endingBefore,
    String? tag,
    String? search,
  }) =>
      _transport.list('/v1/leads', Lead.fromJson, query: {
        'limit': limit,
        'starting_after': startingAfter,
        'ending_before': endingBefore,
        'tag': tag,
        'search': search,
      });

  /// Creates one lead.
  Future<Lead> create(CreateLead body) => _transport
      .request('POST', '/v1/leads', Lead.fromJson, body: body.toJson());

  /// Creates up to 1,000 leads at once. Rows are validated individually;
  /// check `failedCount` and `errors`.
  Future<BulkLeadResult> bulkCreate(BulkCreateLeads body) =>
      _transport.request('POST', '/v1/leads/bulk', BulkLeadResult.fromJson,
          body: body.toJson());

  /// Fetches one lead.
  Future<Lead> retrieve(String id) => _transport.request(
      'GET', '/v1/leads/${Uri.encodeComponent(id)}', Lead.fromJson);

  /// Changes a lead. Only the fields you set are changed.
  Future<Lead> update(String id, UpdateLead body) => _transport.request(
      'PATCH', '/v1/leads/${Uri.encodeComponent(id)}', Lead.fromJson,
      body: body.toJson());

  /// Deletes a lead.
  Future<void> delete(String id) =>
      _transport.requestVoid('DELETE', '/v1/leads/${Uri.encodeComponent(id)}');
}
